core/internal/api/credentials.go

```diff
@@ -55,3 +72,274 @@ func (c *apiKeyCredentialProvider) Apply(req *http.Request) error {
+			return err
+		}
+		c.token = token
+		return nil
+	}
+	if err != nil {
+		c.logger.Warn("failed to access credentials file",
+			"file path", c.credentialsFilePath,
+			"error", err,
+		)
+	}
+
+	token, err := c.loadCredentialsFromFile()
+	if err != nil {
+		return err
+	}
+	c.token = token
+
+	return nil
+}
+
+// loadCredentialsFromFile attempts to load the access token from the credentials file. If
+// the access token does not exist for the given base url, or the token is
+// expiring, it fetches a new one from the server, and saves it to the credentials file
+func (c *oauth2CredentialProvider) loadCredentialsFromFile() (tokenInfo, error) {
+	var credsFile CredentialsFile
+	file, err := os.ReadFile(c.credentialsFilePath)
+	if err != nil {
+		c.logger.Warn("failed to read credentials file",
+			"file path", c.credentialsFilePath,
+			"error", err,
+		)
+	} else {
+		if err := json.Unmarshal(file, &credsFile); err != nil {
+			c.logger.Warn("failed to read credentials file", "error", err.Error())
+		}
+	}
+
+	if credsFile.Credentials == nil {
+		credsFile.Credentials = make(map[string]tokenInfo)
+	}
+
+	creds, ok := credsFile.Credentials[c.baseURL]
+	if !ok || creds.IsTokenExpiring() {
+		creds, err = c.createAccessToken()
+		if err != nil {
+			return tokenInfo{}, err
+		}
+		credsFile.Credentials[c.baseURL] = creds
+		updatedFile, err := json.MarshalIndent(credsFile, "", "  ")
+		if err != nil {
+			c.logger.Warn("failed to update credentials file", "error", err.Error())
+			return creds, nil
+		}
+		err = os.WriteFile(c.credentialsFilePath, updatedFile, 0600)
+		if err != nil {
+			c.logger.Warn("failed to write credentials file", "error", err.Error())
+		}
+	}
+
+	return creds, nil
+}
+
+// writeCredentialsFile fetches an access token from the server and attempts to
+// save it to the credentials file
+func (c *oauth2CredentialProvider) writeCredentialsFile() (tokenInfo, error) {
+	token, err := c.createAccessToken()
+	if err != nil {
+		return tokenInfo{}, err
+	}
+
+	data := CredentialsFile{
+		Credentials: map[string]tokenInfo{
+			c.baseURL: token,
+		},
+	}
+
+	file, err := json.MarshalIndent(data, "", "  ")
+	if err != nil {
+		c.logger.Warn("failed to write credentials file", "error", err.Error())
+		return token, nil
+	}
+
+	err = os.WriteFile(c.credentialsFilePath, file, 0600)
+	if err != nil {
+		c.logger.Warn("failed to write credentials file",
+			"file path", c.credentialsFilePath,
+			"error", err.Error(),
+		)
+	}
+
+	return token, nil
+}
+
+// createAccessToken reads the identity token from a file and exchanges it for
+// an access token from the authorization server using the JWT Bearer flow defined
+// in OAuth 2.0. The access token is then returned with its expiration time.
+func (c *oauth2CredentialProvider) createAccessToken() (tokenInfo, error) {
+	token, err := os.ReadFile(c.identityTokenFilePath)
+	if err != nil {
+		return tokenInfo{}, fmt.Errorf("failed to read identity token file: %v", err)
+	}
```

Comment:
Does the identity token file ever change? Should we be re-reading it here, or should we read the file once on startup? Feels like an unnecessary point of failure, unless we want to allow users to modify their identity token in the middle of a run by changing the file.

Reply:
Good catch, I agree. Moving this to startup